Mixed precision `data_override_mod`

J-Lentz opened the pull request:
Mixed precision support has been added to `data_override_mod`. It has been split into separate implementation and wrapper modules.

**How Has This Been Tested?**
Github CI and unit tests

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

mlee03 commented on data_override/data_override_impl.F90:
```diff
@@ -0,0 +1,2 @@
+#include "data_override_impl_r4.fh"
```
is there a need for this extra file?  It seems to slightly overcomplicate things

J-Lentz replied:
Removed in 28b3d59b56a2a249cff4bfeddca9dc008499153c. Now the implementation modules are included at the top of `data_override.F90` instead.

thomas-robinson commented on data_override/data_override.F90:
```diff
@@ -569,120 +134,93 @@ subroutine data_override_unset_domains(unset_Atm, unset_Ocean, &
-  type(domainUG), intent(inout) :: UGdomain
-  integer, intent(out), optional :: comp_domain(4) !< istart,iend,jstart,jend for compute domain
-  type(domain2D), pointer :: SGdomain => NULL()
+!> @brief Routine to perform data override for scalar fields
```
This description is inaccurate.  This routine selects the correct routine to call to do the data override for scalar fields.

J-Lentz replied:
This subroutine was removed in b89c8db621164eef22e0d9220a2046a4a1ceaa98.

thomas-robinson commented on data_override/data_override.F90:
```diff
@@ -569,120 +134,93 @@ subroutine data_override_unset_domains(unset_Atm, unset_Ocean, &
-!===============================================================================================
+end subroutine data_override_0d
 
 !> @brief This routine performs data override for 2D fields; for usage, see data_override_3d.
```
The description needs to be updated to reflect what this is actually doing

thomas-robinson commented on data_override/data_override.F90:
```diff
@@ -569,120 +134,93 @@ subroutine data_override_unset_domains(unset_Atm, unset_Ocean, &
   logical, intent(out), optional :: override !< true if the field has been overriden succesfully
   type(time_type), intent(in) :: time !<  model time
-  real, dimension(:,:), intent(inout) :: data_2D !< data returned by this call
+  class(*), dimension(:,:), intent(inout) :: data_2D !< data returned by this call
```
This is still not fitting the style of the project using macros.

J-Lentz replied:
This subroutine was removed in b89c8db621164eef22e0d9220a2046a4a1ceaa98.

thomas-robinson commented on data_override/data_override.F90:
```diff
@@ -569,120 +134,93 @@ subroutine data_override_unset_domains(unset_Atm, unset_Ocean, &
+  end select
 end subroutine data_override_2d
 
 !> @brief This routine performs data override for 3D fields
```
The documentation needs to be updated

J-Lentz replied:
This subroutine was removed in b89c8db621164eef22e0d9220a2046a4a1ceaa98.

thomas-robinson commented on data_override/data_override.F90:
```diff
@@ -692,610 +230,53 @@ subroutine data_override_3d(gridname,fieldname_code,data,time,override,data_inde
   type(time_type),              intent(in) :: time !< (target) model time
   integer,           optional,  intent(in) :: data_index
-  real, dimension(:,:,:),    intent(inout) :: data !< data returned by this call
+  class(*), dimension(:,:,:),    intent(inout) :: data !< data returned by this call
```
class(*) should be a macro?

J-Lentz replied:
This subroutine was removed in b89c8db621164eef22e0d9220a2046a4a1ceaa98.

thomas-robinson commented on data_override/data_override.F90:
```diff
@@ -692,610 +230,53 @@ subroutine data_override_3d(gridname,fieldname_code,data,time,override,data_inde
-end subroutine data_override_0d
-
-!> @brief Data override for 2D unstructured grids
+!> @brief Data override for 1D unstructured grids
```
This needs to be updated to reflect what the routine is actually doing

J-Lentz replied:
This subroutine was removed in b89c8db621164eef22e0d9220a2046a4a1ceaa98.

thomas-robinson commented on data_override/include/data_override_impl.inc:
```diff
@@ -191,6 +115,7 @@ subroutine data_override_init(Atm_domain_in, Ocean_domain_in, Ice_domain_in, Lan
   character(len=128)    :: grid_file = 'INPUT/grid_spec.nc'
   integer               :: is,ie,js,je,use_get_grid_version
   integer               :: i, unit, io_status, ierr
+  logical               :: atm_on, ocn_on, lnd_on, ice_on, lndUG_on
```
These variables need documentation

J-Lentz replied:
Documentation for `DATA_OVERRIDE_INIT_IMPL_` arguments added in 2478bf6cf73db4c1126b1b29a50edd75a1deefae.

thomas-robinson commented on data_override/include/data_override_impl.inc:
```diff
@@ -1,188 +1,112 @@
-!! that will override the default values. Real values can be given using data_table, each
-!! line of data_table contains one data_entry. Items of data_entry are comma separated.
-subroutine data_override_init(Atm_domain_in, Ocean_domain_in, Ice_domain_in, Land_domain_in, Land_domainUG_in)
+subroutine DATA_OVERRIDE_INIT_IMPL_(Atm_domain_in, Ocean_domain_in, Ice_domain_in, Land_domain_in, Land_domainUG_in)
```
This routine needs documentation

J-Lentz replied:
Doxygen comment added in 2d6bfd6da5004ab00e7be60e8e85f13cbfa96827.

thomas-robinson commented on data_override/include/data_override_impl.inc:
```diff
@@ -351,8 +274,14 @@ endif
  end if
+end subroutine DATA_OVERRIDE_INIT_IMPL_
+
+function count_ne_1(in_1, in_2, in_3)
```
This function needs documentation

J-Lentz replied:
Doxygen comment added in 59bf15d438dca05b9481e4372e8616b7794c3476.

thomas-robinson commented on data_override/include/data_override_impl.inc:
```diff
@@ -351,8 +274,14 @@ endif
+end subroutine DATA_OVERRIDE_INIT_IMPL_
+
+function count_ne_1(in_1, in_2, in_3)
+  logical, intent(in)  :: in_1, in_2, in_3
+  logical :: count_ne_1
```
these variables need documentation

J-Lentz replied:
Added in 59bf15d438dca05b9481e4372e8616b7794c3476.

thomas-robinson commented on data_override/include/data_override_impl.inc:
```diff
@@ -652,16 +555,102 @@ subroutine get_domainUG(gridname, UGdomain, comp_domain)
+
+end subroutine DATA_OVERRIDE_0D_IMPL_
+
 !> @brief This routine performs data override for 2D fields; for usage, see data_override_3d.
```
I think you want to see data_override_3d_impl_ .  Please update this comment.

J-Lentz replied:
I've removed "for usage, see data_override_3d" in e960b140ae46b8d56cf66737527af978828d62b6.

thomas-robinson commented on data_override/include/data_override_impl.inc:
```diff
@@ -1288,16 +1190,14 @@ subroutine data_override_UG_2d(gridname,fieldname,data,time,override)
-end module data_override_mod
-!> @}
-! close documentation grouping
+end module
```
Please include the module name here

J-Lentz replied:
2b4b16e42a1e2d8fd91257550316b495c79f1a02

thomas-robinson reviewed (changes requested):
1.  Documentation needs to be updated, especially on new routines and new variables.
2. I'm not sure why there are routines that are used to select calling a r4 or r8 routine instead of using an interface.  
3. (or 2.5) My understanding was that unlimited polymorphism wasn't going to be used and instead we were using macros.  This PR uses both.  Please work with @mlee03 to decide how to proceed.  If you do use class(*) because it is needed/neater or whatever, please make sure you document in the code why that was done.

J-Lentz commented on data_override/include/data_override_impl.inc:
```diff
@@ -1,188 +1,112 @@
-use fms2_io_mod,     only : FmsNetcdfFile_t, open_file, close_file, &
-                            read_data, fms2_io_init, variable_exists, &
-                            get_mosaic_tile_file
+module DATA_OVERRIDE_IMPL_
```
As opposed to just calling the macro `DATA_OVERRIDE_`?

J-Lentz commented on data_override/data_override.F90:
```diff
@@ -117,56 +66,18 @@ module data_override_mod
      module procedure data_override_UG_2d
 end interface
 
+integer :: atm_mode = 0
+integer :: ocn_mode = 0
+integer :: lnd_mode = 0
+integer :: ice_mode = 0
```
I added the `class(*)` dispatcher subroutines because I didn't realize that `module procedure x` could be used when `x` is a `use`'d subroutine (I thought `module procedure` was only for procedures in the module's `contains` section). But it actually seems to work fine, so the `class(*)` subroutines have been replaced by interfaces in b89c8db621164eef22e0d9220a2046a4a1ceaa98.

J-Lentz replied:
`atm_mode`, `ocn_mode`, `lnd_mode`, and `ice_mode` all take on values `0` (uninitialized), `4` (r4 mode), and `8` (r8 mode). This will be clarified with doxygen comments.

J-Lentz replied:
Comments added in eb0137c3ee6b3804dea57984bb670168cff9f061.

J-Lentz commented on data_override/data_override.F90:
```diff
@@ -569,120 +134,93 @@ subroutine data_override_unset_domains(unset_Atm, unset_Ocean, &
 
   fail_if_not_set = .true. ; if (present(must_be_set)) fail_if_not_set = must_be_set
 
-  if (.not.module_is_initialized) call mpp_error(FATAL, &
```
1. `module_is_initialized` now lives in the r4 and r8 versions of the implementation modules, rather than in the wrapper module.
2. The new `atm_mode`, `ocn_mode`, `lnd_mode`, and `ice_mode` variables now keep track of what has and hasn't been initialized, so `mpp_error` still gets called if the user tries to unset an uninitialized domain.

J-Lentz commented on data_override/data_override.F90:
```diff
@@ -182,381 +93,35 @@ end function count_ne_1
- if(use_get_grid_version .EQ. 2) then
-   call close_file(fileobj)
- end if
+  select case (mode_selector)
```
We don't receive any real arguments until `data_override` is called. So the only way for `data_override_init` to know the desired kind is for it to be provided as an explicit argument.

J-Lentz commented on data_override/data_override.F90:
```diff
@@ -569,120 +134,93 @@ subroutine data_override_unset_domains(unset_Atm, unset_Ocean, &
+                                                 !! different from the name in NetCDF data file)
+  logical, intent(out), optional :: override !< true if the field has been overriden succesfully
+  type(time_type), intent(in) :: time !< (target) model time
+  class(*),             intent(out) :: data !< output data array returned by this call
```
This subroutine was removed in b89c8db621164eef22e0d9220a2046a4a1ceaa98.

J-Lentz commented on data_override/data_override.F90:
```diff
@@ -569,120 +134,93 @@ subroutine data_override_unset_domains(unset_Atm, unset_Ocean, &
+                                                 !! different from the name in NetCDF data file)
+  logical, intent(out), optional :: override !< true if the field has been overriden succesfully
+  type(time_type), intent(in) :: time !< (target) model time
+  class(*),             intent(out) :: data !< output data array returned by this call
```
This subroutine was removed in b89c8db621164eef22e0d9220a2046a4a1ceaa98.

J-Lentz commented on data_override/data_override.F90:
```diff
@@ -182,381 +93,35 @@ end function count_ne_1
   type (domain2d), intent(in), optional :: Land_domain_in
   type(domainUG) , intent(in), optional :: Land_domainUG_in
+  integer, intent(in), optional :: mode !< r4_kind or r8_kind
+  integer :: mode_selector
```
Meatier comment added in eb0137c3ee6b3804dea57984bb670168cff9f061.

J-Lentz commented on data_override/data_override.F90:
```diff
@@ -182,381 +93,35 @@ end function count_ne_1
 !! that will override the default values. Real values can be given using data_table, each
 !! line of data_table contains one data_entry. Items of data_entry are comma separated.
-subroutine data_override_init(Atm_domain_in, Ocean_domain_in, Ice_domain_in, Land_domain_in, Land_domainUG_in)
+subroutine data_override_init(Atm_domain_in, Ocean_domain_in, Ice_domain_in, Land_domain_in, Land_domainUG_in, mode)
```
Updated in e960b140ae46b8d56cf66737527af978828d62b6.

J-Lentz commented on data_override/data_override.F90:
```diff
@@ -569,120 +134,93 @@ subroutine data_override_unset_domains(unset_Atm, unset_Ocean, &
+  if (present(unset_Atm)) then ; if (unset_Atm) then
+    select case (atm_mode)
+      case (r4_kind)
+        call data_override_unset_atm_r4
+      case (r8_kind)
+        call data_override_unset_atm_r8
```
Which compiler was it an issue with? It seems to work with ifort, ifx, gfortran, nvfortran, and Cray.

mlee03 commented:
cool :)  For consistency, could you change DATA_OVERRIDE_KIND_ to start with FMS_DATA_OVERRIDE_KIND_ (or something like that that starts with FMS).  Same for GET_GRID_VERSION_KIND_

J-Lentz commented:
Macro names changed in 2b71679810237aca8455745ecacbc998e20f1e37.

rem1776 commented on data_override/data_override.F90:
```diff
@@ -37,1267 +41,163 @@
-                            get_mosaic_tile_file
-use get_grid_version_mod, only: get_grid_version_1, get_grid_version_2
+  use data_override_r4
+  use data_override_r8
```
So all the derived types are private and named the same within these two modules right?

J-Lentz replied:
That's correct - only the public subroutines have different `r4`/`r8` names.

rem1776 reviewed (changes requested):
I don't think we combine multiple modules in the same file elsewhere in fms, do you know if it gives the right line numbers in tracebacks for includes if you got an error from one of those r4/r8 routines?

Creating two more files just to include the two kind versions doesn't make much more sense either though, so I would add some comments since this switches up the typical mixed precision CPP method. A line or two in the header comments for data_override.F90 and data_override.inc explaining how this include is used as a module in this case specifically would be good.

J-Lentz commented on data_override/Makefile.am:
```diff
@@ -23,19 +23,26 @@
 
 # Some mods are dependent on other mods in this dir.
 data_override_mod.$(FC_MODEXT): get_grid_version_mod.$(FC_MODEXT)
+data_override_mod.$(FC_MODEXT): get_grid_version_mod.$(FC_MODEXT)
```
Yep, careless mistake. Fixed in ca0d5ab2187bcadcfc111a41ebfccf84e96e6cdc.

mlee03 commented:
@rem1776 @J-Lentz, I can see how using FMS_DATA_OVERRIDE_TYPE and explicitly declaring two versions of the private types (as done in the rest of mixed-precision FMS) could work, but I like how data_override is modified here, it's very clean.  And if we in the future add submodules, the changes would be straightforward here.  And agree with @rem1776, I now see how the mixed precision implemented here is not the easiest to understand and a bit of comment would be beneficial.

J-Lentz commented:
I've added brief comments to the tops of `data_override.F90` and `data_override.inc`. And I've moved the `module data_override_r*` and `module end data_override_r*` statements to `data_override.F90`, because Fortran code in .fh files could be confusing and unexpected.

mlee03 commented on data_override/include/data_override.inc:
```diff
@@ -871,8 +907,8 @@ subroutine data_override_3d(gridname,fieldname_code,data,time,override,data_inde
 ! convert lon_in and lat_in from deg to radian
-        override_array(curr_position)%lon_in = override_array(curr_position)%lon_in * deg_to_radian
-        override_array(curr_position)%lat_in = override_array(curr_position)%lat_in * deg_to_radian
+        override_array(curr_position)%lon_in = override_array(curr_position)%lon_in * DEG_TO_RAD
```
real(DEG_TO_RAD,lkind)

mlee03 commented on test_fms/data_override/test_get_grid_v1.F90:
```diff
@@ -129,11 +131,11 @@ program test_get_grid_v1
 
 !< Error checking:
-if (lon(1,1) .ne. sum(lon_vert_in)/4*deg_to_radian ) then
+if (lon(1,1) .ne. sum(lon_vert_in)/4*real(DEG_TO_RAD, lkind) ) then
```
4.0_lkind

J-Lentz commented on data_override/get_grid_version.F90:
```diff
@@ -33,7 +34,16 @@ module get_grid_version_mod
 
 implicit none
 
-real, parameter    :: deg_to_radian=PI/180.
```
It was, but I've replaced it with `DEG_TO_RAD` from `constants_mod`.
